From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Alexey Makhalov <alexey.makhalov@broadcom.com>
Date: Mon, 6 Apr 2026 00:00:00 +0000
Subject: [PATCH] domain: add domain_use_openssl_alg_socket for AF_ALG (FIPS)

Introduce attribute openssl_alg_socket_domain and interface
domain_use_openssl_alg_socket() so OpenSSL-using daemons (FIPS
jitterentropy via AF_ALG) share one allow rule instead of per-domain
allows.

Use it for syslogd_t (rsyslog TLS), sshd_t, and ssh_keygen_t.  Replaces
direct alg_socket allows from 0035-allow-alg_socket-for-sshd.patch for
ssh; the redundant second sshd_t alg_socket allow at the end of ssh.te
is removed (single domain_use_openssl_alg_socket(sshd_t) earlier).

---
 policy/modules/kernel/domain.if     | 26 ++++++++++++++++++++++++++
 policy/modules/kernel/domain.te     |  5 +++++
 policy/modules/services/ssh.te      |  8 +++-----
 policy/modules/system/logging.te    |  2 ++
 4 files changed, 36 insertions(+), 5 deletions(-)

diff --git a/policy/modules/kernel/domain.if b/policy/modules/kernel/domain.if
index 0e84372..eb92889 100644
--- a/policy/modules/kernel/domain.if
+++ b/policy/modules/kernel/domain.if
@@ -1842,3 +1842,29 @@ interface(`domain_rw_perf_event_all_domains',`
 
 	allow $1 domain:perf_event rw_inherited_perf_event_perms;
 ')
+
+########################################
+## <summary>
+##	Allow domain to use AF_ALG sockets (OpenSSL FIPS / jitterentropy).
+## </summary>
+## <desc>
+##	<p>
+##	On Photon OS with FIPS enabled, OpenSSL may use the kernel AF_ALG
+##	interface (e.g. jitterentropy).  Assign the process type to attribute
+##	openssl_alg_socket_domain so it receives the common self:alg_socket
+##	allow rule defined in domain.te.
+##	</p>
+## </desc>
+## <param name="domain">
+##	<summary>
+##	Domain type allowed access (e.g. syslogd_t, sshd_t).
+##	</summary>
+## </param>
+#
+interface(`domain_use_openssl_alg_socket',`
+	gen_require(`
+		attribute openssl_alg_socket_domain;
+	')
+
+	typeattribute $1 openssl_alg_socket_domain;
+')
diff --git a/policy/modules/kernel/domain.te b/policy/modules/kernel/domain.te
index 72ffcd8..928fb83 100644
--- a/policy/modules/kernel/domain.te
+++ b/policy/modules/kernel/domain.te
@@ -55,6 +55,9 @@ gen_tunable(domain_can_mmap_files, false)
 attribute domain;
 attribute named_filetrans_domain;
 
+# Domains that use OpenSSL with AF_ALG (FIPS jitterentropy on Photon, etc.)
+attribute openssl_alg_socket_domain;
+
 # Transitions only allowed from domains to other domains
 neverallow domain ~domain:process { transition dyntransition };
 
@@ -130,6 +133,8 @@ allow domain self:fifo_file rw_fifo_file_perms;
 allow domain self:sem create_sem_perms;
 allow domain self:shm create_shm_perms;
 
+allow openssl_alg_socket_domain self:alg_socket { create bind accept read write getattr setopt };
+
 # This is a temporary rule to work around a problem in kernel/xfs
 # triggering a false fowner capability AVC
 # https://bugzilla.redhat.com/show_bug.cgi?id=1933437
diff --git a/policy/modules/services/ssh.te b/policy/modules/services/ssh.te
index 7e46799..2ca5344 100644
--- a/policy/modules/services/ssh.te
+++ b/policy/modules/services/ssh.te
@@ -47,7 +47,7 @@ type ssh_keygen_t;
 type ssh_keygen_exec_t;
 init_system_domain(ssh_keygen_t, ssh_keygen_exec_t)
 
-allow ssh_keygen_t self:alg_socket { create bind accept read write getattr setopt };
+domain_use_openssl_alg_socket(ssh_keygen_t)
 
 type ssh_keygen_tmp_t;
 files_tmp_file(ssh_keygen_tmp_t)
@@ -320,7 +320,7 @@ term_setattr_all_ttys(sshd_t)
 term_relabelto_all_ptys(sshd_t)
 term_use_ptmx(sshd_t)
 
-allow sshd_t self:alg_socket { create bind accept };
+domain_use_openssl_alg_socket(sshd_t)
 motdgen_exec(sshd_t);
 
 # for X forwarding
@@ -748,10 +748,8 @@ optional_policy(`
 allow sshd_t var_log_t:file { create open read write };
 allow sshd_t var_log_t:file { setattr };
 
-allow sshd_t self:alg_socket { create bind accept read write getattr setopt };
-
 # When sshd forks unix_chkpwd (transitioning sshd_t -> chkpwd_t via execve),
 # the kernel checks siginh/rlimitinh/noatsecure on the new domain.
 # Without these rules the child starts with reset signal dispositions and
 # resource limits, and AT_SECURE=1 (glibc secure mode).  Authentication
 # still works but three AVCs are logged on every SSH login attempt:
 #   avc: denied { siginh } for comm="unix_chkpwd"
 #     scontext=system_u:system_r:sshd_t tcontext=system_u:system_r:chkpwd_t
 #     tclass=process permissive=0
 allow sshd_t chkpwd_t:process { noatsecure rlimitinh siginh };
diff --git a/policy/modules/system/logging.te b/policy/modules/system/logging.te
index 6ab5f77..fac2b2c 100644
--- a/policy/modules/system/logging.te
+++ b/policy/modules/system/logging.te
@@ -111,6 +111,8 @@ init_daemon_domain(syslogd_t, syslogd_exec_t)
 init_nnp_daemon_domain(syslogd_t)
 mls_trusted_object(syslogd_t)
 
+domain_use_openssl_alg_socket(syslogd_t)
+
 libs_read_lib_files(syslogd_t)
 
 type syslogd_initrc_exec_t;
